crates/sparrow-scheduler/src/worker.rs

```diff
@@ -0,0 +1,74 @@
+use crate::{queue::*, Error, TaskRef};
+
+/// An injector that allows adding work to the global queue.
+#[derive(Debug, Clone)]
+pub struct Injector {
+    queue: GlobalQueue<TaskRef>,
+}
+
+impl Injector {
+    pub fn create(workers: usize, local_queue_size: u16) -> (Self, Vec<Worker>) {
+        let queue = GlobalQueue::new(workers, local_queue_size);
+        let workers = queue
+            .take_local_queues()
+            .map(|queue| Worker { queue })
+            .collect();
+        (Injector { queue }, workers)
+    }
+}
+
+impl Queue<TaskRef> for Injector {
+    fn push_global(&self, task: TaskRef) {
+        if task.schedule() {
+            self.queue.push_global(task)
+        }
+    }
+
+    fn push(&mut self, task: TaskRef) {
+        self.push_global(task)
+    }
+
+    fn push_yield(&mut self, task: TaskRef) {
+        self.push_global(task)
+    }
+}
+
+/// An individual worker that allows adding work to the local or global queue.
+pub struct Worker {
+    queue: LocalQueue<TaskRef>,
+}
+
+impl Worker {
+    /// Run the work loop to completion.
+    pub(crate) fn work_loop(mut self) -> error_stack::Result<(), Error> {
+        while let Some(task) = self.queue.pop() {
```

Comment:
I think this needs some added logic. Specifically, we should probably sleep if there are no tasks and wait for new work to be injected rather than exiting the work loop.

Reply:
What is in the "work loop" Does a worker handle work for one pipeline? Or when _should_ it exit?

Reply:
Workers take tasks from their queue and run them. Remember that with morsel driven parallelism we prefer one thread focusing on a specific batch of data and running *all pipelines*. This is actually why the pipelines themselves need mutexes and things in the pipelines so that we can make sure that when a thread starts working on it it has exclusive rights to that.